### PR TITLE
Fix appname install directory

### PR DESCRIPTION
**Pull request by phantomtypist**
Paul, I think I fixed the problem where the directory name of the installed app was not getting set properly if a package Id had a period in it.

Please review :worried:

### CODE REVIEW

File: src/Squirrel/ReleaseEntry.cs [hunk 1]
```diff
@@ -53,11 +54,24 @@ protected ReleaseEntry(string sha1, string filename, long filesize, bool isDelta
         [IgnoreDataMember]
         public Version Version { get { return Filename.ToVersion(); } }
 
+        [Obsolete("Property is deprecated because it may not correctly retrieve package names (a.k.a. Id's) if you have periods in it.  If you wish to retrieve the package Id, please use the GetPackageId method.")]
         [IgnoreDataMember]
         public string PackageName {
             get { return Filename.Substring(0, Filename.IndexOfAny(new[] { '-', '.' })); }
```
Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Why don't you just fix this method to correctly handle IDs with '.'s in it?

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: I was a little worried about breaking functionality for people that might be using it and relying on it to work the way it was working.

Another thought was I didn't want to make any more assumptions about parsing a filename to find the package Id/Name... too many unknowns to worry about.  I figured it was just best to go to the source of where it is inside the NuGet package metadata...  kinda like how you are getting the release notes.

My last thought was that "package Id" was a better fit than "package name".  Now, I know that in the description of the "package Id" in the Nuget nuspec, they refer to the Id as the name of the package.  However, I've run into people in the past that have gotten confused and associated a "package name" with the actual title of the package.

Let me know if you want me to make some modifications.

Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: > I was a little worried about breaking functionality for people that might be using that property and relying on it to work the way it was working.

This is a good impulse but this API isn't *actually* intended for anyone to be used, and at the moment, it is returning incorrect results (i.e. nobody can be relying on it returning an incorrect package name because it's not returning anything useful for those people)

> Another thought was I didn't want to make any more assumptions about parsing a filename to find the package Id/Name... too many unknowns to worry about.

This is a fundamental part of Release Entry, that the name *must be* of the format "PackageName-Version-[delta/full].nupkg". This is because when we read the RELEASES file, we don't have access to the NuGet package, so we can't read the metadata.

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: Alright, let me take another crack at it.

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: My one thought was the format "PackageName-Version-[delta/full].nupkg" could be hard to parse. You could have package names with periods and hyphens... and also at the same time the version could be semantic with hyphens too (e.g. 1.1.0.0-beta.)

Would you be opposed to me using a regular expression?

Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: > You could have package names with periods and hyphens

Yeah, probably easiest to start at the back and parse towards the front. 

> and also at the same time the version could be semantic with hyphens too (e.g. 1.1.0.0-beta.)

Prerelease tags aren't supported, ignore that case

> Would you be opposed to me using a regular expression?

Not at all :)

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: > Prerelease tags aren't supported, ignore that case

Can you help me understand why prerelease tags aren't supported?

Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Because we use the "slot" for prerelease tags to tag as "delta" or "full" package

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: > Because we use the "slot" for prerelease tags to tag as "delta" or "full" package

I get it now.  I was in the mindset thinking the "delta" and "full" tags were extra stuff outside of the version scope.

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: Since prerelease tags aren't supported, is there an actual way/strategy to release prerelease applications (i.e. beta)?

Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Use a different release channel - Squirrel was built s.t. you can easily set up Channels a-la Chrome / Firefox, where different users get different sets of updates, just by changing the folder / URL you update from

**Comment by phantomtypist** [comment]
@paulcbetts FYI I'm done making changes.  Ready for review.

File: src/Squirrel/ReleaseEntry.cs [hunk 2]
```diff
@@ -53,9 +53,19 @@ protected ReleaseEntry(string sha1, string filename, long filesize, bool isDelta
         [IgnoreDataMember]
         public Version Version { get { return Filename.ToVersion(); } }
 
+        static readonly Regex packageRegex = new Regex(@"^(?<packageName>.*)-(?<version>\d+\.\d+\.\d+\.\d+)-(?<releaseType>full|delta).nupkg$");
         [IgnoreDataMember]
         public string PackageName {
-            get { return Filename.Substring(0, Filename.IndexOfAny(new[] { '-', '.' })); }
+            get
+            {
```
Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Move this bracket on the previous line

File: src/Squirrel/ReleaseEntry.cs [hunk 3]
```diff
@@ -53,9 +53,19 @@ protected ReleaseEntry(string sha1, string filename, long filesize, bool isDelta
...
+                var match = packageRegex.Match(Filename);
+
+                if (!match.Success) {
```
Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Can you move this validation into the place where we provide this information (i.e. the ctor?) Getters that throw are evil

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: It's gonna break a lot of the tests because the InlineData for the test has invalid package names in certain ones.

Do you mind if I change those as well?

Comment (phantomtypist) [inline src/Squirrel/ReleaseEntry.cs]: Every filename in the release entry will always contains "-delta|-full" at the end before the ".nupkg" extension?

Comment (paulcbetts) [inline src/Squirrel/ReleaseEntry.cs]: Yes, that's guaranteed

**Comment by JakeGinnivan** [comment]
Would love this to be merged, would have saved me a bit of time.

**Comment by Pjanssen** [comment]
Is there any chance of this PR being merged? 
I'm running into issues where I have multiple applications with a package ID in this format: `CompanyName.Product.SubProduct`. All of these subproducts are getting installed into the same directory `CompanyName`.

**Comment by phantomtypist** [comment]
@Pjanssen I'm busy with life at the moment, but I'll see if I can find time in the near future, unless you want to finish it for me?

**Comment by Pjanssen** [comment]
@phantomtypist Could you describe what is left to be done?
